tests/test_argparcel.py

```diff
@@ -193,3 +198,61 @@ def test_missing_argument() -> None:
         argparse.ArgumentError, match="arguments are required: --a/--no-a"
     ):
         argparcel.parse(ArgsRequiredFlag, [], exit_on_error=False)
+
+
+@dataclasses.dataclass
+class BadUnderscore1:
+    _a: bool
+
+
+@dataclasses.dataclass
+class BadUnderscore2:
+    # pyright tells us we're not allowed to do this. For the sake of our testing, let's
+    # brute force it.
+    __a: bool  # pyright: ignore [reportGeneralTypeIssues]
+
+    def moo(self) -> bool:
+        # This exists purely so pyright doesn't see `__a` as being unaccessed.
+        return self.__a
+
+
+def test_bad_underscores() -> None:
+    with pytest.raises(
+        ValueError, match="Field names must not start with an underscore; got '_a'"
+    ):
+        argparcel.parse(BadUnderscore1, ["--help"])
+
+    with pytest.raises(
+        ValueError,
+        match="Field names must not start with an underscore; got '_BadUnderscore2__a'",
+    ):
+        argparcel.parse(BadUnderscore2, ["--help"])
+
+
+@dataclasses.dataclass
+class MooWithMethods:
+    a: bool
+
+    @property
+    def b(self) -> bool:
+        return self.a
+
+    def c(self) -> bool:
+        return self.a
+
+
+def test_properties_and_methods_ok() -> None:
```

Comment:
Additional test unrelated to the issue that this PR closes, but it came to mind.